REGRESSION (262176@main): Don't recreate the ScrollbarsController every time we add a scrollbar

smfr opened the pull request:
#### 003e0fd9799fd84d392a22db6bcf7f4c2f8ddf0d
<pre>
REGRESSION (262176@main): Don&apos;t recreate the ScrollbarsController every time we add a scrollbar
<a href="https://bugs.webkit.org/show_bug.cgi?id=257399">https://bugs.webkit.org/show_bug.cgi?id=257399</a>
rdar://109909019

Reviewed by Tim Nguyen.

Rather than calling setScrollbarsController() on every call to LocalFrameView::didAddScrollbar(),
refactor the factory functions so that the clients (LocalFrameView and. RenderLayerScrollableArea)
can make the right kind of ScrollbarsController via a virtual function, and only call this
factory function once.

This also ensures that we make a ScrollbarsControllerMock when we need to, for testing.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::createScrollbarsController):
(WebCore::LocalFrameView::didAddScrollbar):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::scrollbarsController const):
(WebCore::ScrollableArea::internalCreateScrollbarsController):
(WebCore::ScrollableArea::createScrollbarsController):
* Source/WebCore/platform/ScrollableArea.h:
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::createScrollbarsController):
(WebCore::RenderLayerScrollableArea::createScrollbar):
* Source/WebCore/rendering/RenderLayerScrollableArea.h:

Canonical link: <a href="https://commits.webkit.org/264690@main">https://commits.webkit.org/264690@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c8696e2d50661bce9c634b86ee634753e28d1238

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8189 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8481 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8698 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9853 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8258 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8197 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10469 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8392 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11133 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8334 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9400 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7431 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9975 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6700 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7494 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15055 "2 flakes 114 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7823 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7624 "2 failures") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10986 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8096 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6580 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7387 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2011 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11594 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7837 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -5813,14 +5813,21 @@ void LocalFrameView::setScrollingPerformanceTestingEnabled(bool scrollingPerform
-        if (auto scrollbarController = page->chrome().client().createScrollbarsController(*page, *this))
-            setScrollbarsController(WTFMove(scrollbarController));
+    auto* page = m_frame->page();
+    if (auto scrollbarController = page->chrome().client().createScrollbarsController(*page, *this)) {
```
What guarantees that page isn't null?

smfr committed:
REGRESSION (262176@main): Don't recreate the ScrollbarsController every time we add a scrollbar
https://bugs.webkit.org/show_bug.cgi?id=257399
rdar://109909019

Reviewed by Tim Nguyen.

Rather than calling setScrollbarsController() on every call to LocalFrameView::didAddScrollbar(),
refactor the factory functions so that the clients (LocalFrameView and. RenderLayerScrollableArea)
can make the right kind of ScrollbarsController via a virtual function, and only call this
factory function once.

This also ensures that we make a ScrollbarsControllerMock when we need to, for testing.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::createScrollbarsController):
(WebCore::LocalFrameView::didAddScrollbar):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::scrollbarsController const):
(WebCore::ScrollableArea::internalCreateScrollbarsController):
(WebCore::ScrollableArea::createScrollbarsController):
* Source/WebCore/platform/ScrollableArea.h:
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::createScrollbarsController):
(WebCore::RenderLayerScrollableArea::createScrollbar):
* Source/WebCore/rendering/RenderLayerScrollableArea.h:

Canonical link: https://commits.webkit.org/264690@main
